File: frontend/src/types/report/report.ts
/**
 * Report JSW/JVML — domain types mirroring the backend `report` domain
 * (`schemas/report.py`). RAKE-pivot layout: rows grouped by SO Sales Org →
 * Distr. Channel → BRANCH (Sales Office) → Sold To Party → Party Code →
 * Ship-To Party → Transport Mode → Destination → ROUTE; dynamic RAKE columns
 * contain the sum of stock_quantity. Credit-report checks are preserved.
 */

export type ReportType = "jsw" | "jvml"
/** Toolbar selection — `both` fans out to two API calls (jsw + jvml), grouped by SO Sales Org. */
export type ReportTypeSelection = ReportType | "both"
// QA-hold aging day-filter ("aged QA-hold" = in_quality_insp > 0 AND round(aging) > 2):
//   include = all stock · exclude = all EXCEPT aged QA-hold · only = ONLY aged QA-hold
export type DaysFilter = "include" | "exclude" | "only"
export type CreditStatus = "" | "No Credit balance" | "NO CREDIT REPORT FOUND"

/** One row of the RAKE pivot report. */
export interface ReportPivotRow {
  so_sales_org: string | null
  distr_chnl: string | null
  sold_to_party: string | null
  sales_office: string | null      // rendered as BRANCH
  party_code: string               // normalized display code
  ship_to_party: string | null
  customer_name: string | null     // stock-row customer; for the 8-field exclusion identity
  transport_mode: string | null    // from CustomerCode.transport_mode
  destination: string | null       // from CustomerCode.destination
  route: string | null             // from CustomerCode.route
  rake_quantities: Record<string, number>  // { "KKU": 55.01, "ROAD": 0, ... }
  total: number                    // Σ stock_quantity for this row
  nco_yes_do: number               // Σ stock_quantity where NCO Declared == "Yes"
  nco_yes_do_count: number
  blocked: boolean | null          // null = no credit data
  credit_balance: number | null    // SAP headroom; null = not in credit report
  required_credit: number | null   // (total - nco_yes_do) * coil_price_per_qty
  credit_sufficient: boolean | null
  credit_status: CreditStatus
  credit_note: string              // Balance Available / Not Enough Balance / Balance Negative / …
}

/** Full report payload (the `data` inside the success envelope). */
export interface ReportResponse {
  date: string
  report_type: ReportTypeSelection  // "both" ⇒ merged jsw + jvml, grouped by SO Sales Org
  region_id: string | null
  region_name: string
  days_filter: DaysFilter
  ccas: string[]                    // e.g. ["VJ0H", "1000"] (jsw) / ["JV0H"] (jvml)
  has_stock: boolean                // false → "No stock excel for this date selected"
  has_credit_report: boolean        // false → credit columns "NO CREDIT REPORT FOUND"
  coil_price_per_qty: number | null
  rake_columns: string[]            // sorted unique RAKE values for this report
  rows: ReportPivotRow[]
  grand_total: number
  grand_nco_yes_do: number
  grand_required_credit: number | null
  rake_totals: Record<string, number>            // RAKE → Σ stock_quantity
  transport_mode_totals: Record<string, number>  // transport_mode → Σ stock_quantity
}

/** One individual stock row contributing to a RAKE total (drill-down). */
export interface RakeDrilldownRow {
  stock_type: ReportType           // "jsw" | "jvml" — which collection it came from
  so_sales_org: string | null      // Sales Org
  distr_chnl: string | null        // Distr Channel
  sold_to_party: string | null     // Sold to party
  sales_office: string | null      // rendered as BRANCH
  party_code: string | null        // normalized display code
  ship_to_party: string | null     // Ship to party
  transport_mode: string | null    // from CustomerCode.transport_mode
  destination: string | null       // from CustomerCode.destination
  customer_name: string | null     // mapped customer
  stock_quantity: number           // this row's quantity
}

/** A merged drill-down row: 8-field identity + summed quantity (Source & Ship To Party dropped). */
export interface RakeDrilldownMergedRow {
  so_sales_org: string | null      // Sales Org
  distr_chnl: string | null        // Distr Channel
  sales_office: string | null      // rendered as BRANCH
  sold_to_party: string | null     // Sold to party
  party_code: string | null        // normalized display code
  transport_mode: string | null    // from CustomerCode.transport_mode
  destination: string | null       // from CustomerCode.destination
  customer_name: string | null     // mapped customer
  stock_quantity: number           // Σ stock_quantity for the merged group
}

/** GET /report/rake-drilldown payload — individual jsw + jvml rows for one RAKE. */
export interface RakeDrilldownResponse {
  rake: string
  date: string
  region_id: string | null
  region_name: string
  days_filter: DaysFilter
  rows: RakeDrilldownRow[]
  merged_rows: RakeDrilldownMergedRow[]   // rows collapsed by 8-field identity (server-computed)
  total_quantity: number
}

/** Query params for GET /report/rake-drilldown. */
export interface RakeDrilldownParams {
  rake: string
  date: string
  region_id?: string
  days: DaysFilter
}

/** Query params for GET /report/generate (date is "dd-MM-yyyy"). */
export interface ReportQueryParams {
  date: string
  report_type: ReportTypeSelection  // "both" ⇒ one merged call (jsw + jvml)
  region_id?: string
  days: DaysFilter
  columns?: string                  // export only: CSV of visible optional-column keys
}

/** Which sheets the combined /report export should include (picker dialog keys). */
export type ExportSheetKey =
  | "pivot"          // BRANCH WISE PIVOT REPORT
  | "rake_totals"    // TOTAL RAKE REPORT
  | "rake_merged"    // one "{RAKE} - Total Rake Wise" sheet per rake (merged rows)
  | "rake_unmerged"  // one "{RAKE} - Batch Rake Wise" sheet per rake (raw rows)
  | "jsw"            // JSW Stock List
  | "jvml"           // JVML Stock List
  | "credit"         // Credit Report

/** One RAKE's browser-only exclusions, sent as the export POST body (per rake). */
export interface RakeExclusionWire {
  keys: string[]      // canonical row_identity strings to omit from breakdown sheets
  subtract: number    // qty to subtract from this rake's TOTAL RAKE REPORT figure
}

/** Query params for GET|POST /report/export-combined (report params + chosen sheets). */
export interface CombinedExportParams extends ReportQueryParams {
  sheets: ExportSheetKey[]
  /** Browser-only RAKE drill-down exclusions; present ⇒ POST with a JSON body. */
  exclusions?: Record<string, RakeExclusionWire>
  /** Same unchecks regrouped by transport mode → subtract from the transport-mode totals sheet. */
  transport_subtract?: Record<string, number>
}

File: backend/app/schemas/report.py
"""Report JSW/JVML request/response DTOs.

Contract for ``GET /report/generate`` — a "Coil Stock" RAKE pivot for a date +
report type + region. Stock rows are grouped by SO Sales Org → Distr. Channel →
Sold To Party → Sales Office (BRANCH) → Party Code → Ship-To Party, then
enriched with CustomerCode master fields (Transport Mode, Destination, ROUTE).
RAKE values from CustomerCode become dynamic column headers; the value is the
sum of stock_quantity for that RAKE. Credit-report checks are kept unchanged.

See ``.planning/report-jsw-jvml/SPEC.md`` for the full algorithm.
"""

from __future__ import annotations

from pydantic import BaseModel, Field
from typing import Literal

# Report type → stock collection + credit control areas:
#   jsw  → jsw_stock  + CCAs ["VJ0H", "1000"]
#   jvml → jvml_stock + CCA ["JV0H"]
ReportType = Literal["jsw", "jvml"]
# Toolbar/query selection. "both" merges jsw + jvml into one response, grouped by
# SO Sales Org (each stock model is still queried as its own ReportType internally).
ReportTypeSelection = Literal["jsw", "jvml", "both"]

# QA-hold aging day-filter, defined around "aged QA-hold" stock =
# in_quality_insp > 0 AND round(aging) > 2 (aging rounded to whole days, 2.4→2,
# 2.6→3). Normal non-QA stock is never "aged QA-hold".
#   include → no filter (all stock)
#   exclude → everything EXCEPT aged QA-hold (all normal data + QA-hold ≤ 2 days)
#   only    → ONLY aged QA-hold (in_quality_insp > 0 AND round(aging) > 2)
DaysFilter = Literal["include", "exclude", "only"]

# Credit-status sentinels rendered verbatim by the frontend.
CreditStatus = Literal["", "No Credit balance", "NO CREDIT REPORT FOUND"]


class ReportQuery(BaseModel):
    """Query params for ``GET /report/generate`` (used via FastAPI ``Depends()``)."""

    date: str = Field(pattern=r"^\d{2}-\d{2}-\d{4}$")
    report_type: ReportTypeSelection
    region_id: str | None = Field(default=None, max_length=64)
    days: DaysFilter = "include"
    # Export only: CSV of visible optional-column keys (e.g. "total,route,blocked").
    # None ⇒ every optional column; "" ⇒ none. Ignored by /generate.
    columns: str | None = Field(default=None, max_length=256)


class CombinedExportQuery(ReportQuery):
    """Query params for ``GET /report/export-combined`` (via FastAPI ``Depends()``).

    The report export params plus ``sheets`` — a CSV of which sheets to include
    (subset of pivot, rake_totals, rake_merged, rake_unmerged, jsw, jvml, credit).
    Value validation (unknown / empty keys) happens in the combined service.
    """

    sheets: str = Field(min_length=1, max_length=128)


class RakeExclusion(BaseModel):
    """One RAKE's user-unchecked drill-down rows (export-only, transient).

    ``keys`` — canonical 8-field identity strings (see ``rake_drilldown.row_identity``,
    mirrored by the frontend ``rake-exclusions.ts::rowKey``) of rows to OMIT from the
    rake_merged / rake_unmerged breakdown sheets. ``subtract`` — quantity to subtract
    from this RAKE's TOTAL RAKE REPORT figure (stock-type-scoped, computed client-side).
    Nothing is persisted; this only shapes one export.
    """

    keys: list[str] = Field(default_factory=list, max_length=10000)
    subtract: float = 0.0


class CombinedExportBody(BaseModel):
    """Optional POST body for ``/report/export-combined`` — per-export RAKE exclusions.

    ``exclusions`` maps RAKE → its excluded rows (subtract from the RAKE totals +
    drop from the breakdown sheets). ``transport_subtract`` maps transport mode →
    the same unchecks' qty (subtract from the TRANSPORT MODE TOTAL sheet). Both
    empty/absent ⇒ no exclusions (GET-equivalent file).
    """

    exclusions: dict[str, RakeExclusion] = Field(default_factory=dict)
    transport_subtract: dict[str, float] = Field(default_factory=dict)


class ReportPivotRow(BaseModel):
    """One row of the RAKE pivot report."""

    so_sales_org: str | None
    distr_chnl: str | None
    sold_to_party: str | None
    sales_office: str | None       # rendered as BRANCH
    party_code: str                # normalized display code
    ship_to_party: str | None
    transport_mode: str | None     # from CustomerCode.transport_mode
    destination: str | None        # from CustomerCode.destination
    route: str | None              # from CustomerCode.route
    rake_quantities: dict[str, float] = {}  # { "KKU": 55.01, "ROAD": 0, ... }
    total: float                   # Σ stock_quantity for this row
    nco_yes_do: float              # Σ stock_quantity where nco_declared == "Yes"
    nco_yes_do_count: int          # # rows where nco_declared == "Yes"
    blocked: bool | None           # credit Blocked flag; None = no credit data
    credit_balance: float | None   # SAP credit headroom; None = not in credit report
    required_credit: float | None  # (total - nco_yes_do) * coil_price_per_qty
    credit_sufficient: bool | None # credit_balance >= required_credit (None if either missing)
    credit_status: CreditStatus    # "" | "No Credit balance" | "NO CREDIT REPORT FOUND"
    credit_note: str               # verdict label


class RakeDrilldownQuery(BaseModel):
    """Query params for ``GET /report/rake-drilldown`` (via FastAPI ``Depends()``).

    Reverse-resolves a single RAKE to the individual jsw + jvml stock rows that
    make up its total. Always queries BOTH stock collections (the report's
    ``report_type`` is irrelevant here — the drill-down is union jsw + jvml).
    """

    rake: str = Field(min_length=1, max_length=64)
    date: str = Field(pattern=r"^\d{2}-\d{2}-\d{4}$")
    region_id: str | None = Field(default=None, max_length=64)
    days: DaysFilter = "include"


class RakeDrilldownRow(BaseModel):
    """One individual stock row contributing to a RAKE total."""

    stock_type: ReportType         # "jsw" | "jvml" — which collection it came from
    so_sales_org: str | None       # Sales Org
    distr_chnl: str | None         # Distr Channel
    sold_to_party: str | None      # Sold to party
    sales_office: str | None       # rendered as BRANCH
    party_code: str | None         # normalized display code
    ship_to_party: str | None      # Ship to party
    transport_mode: str | None     # from CustomerCode.transport_mode
    destination: str | None        # from CustomerCode.destination
    customer_name: str | None      # mapped customer (for readability)
    stock_quantity: float          # this row's quantity


class RakeDrilldownMergedRow(BaseModel):
    """A merged drill-down row: the 8-field business identity + summed quantity.

    Source (stock_type) and Ship To Party are intentionally absent — they are not
    part of the merge key and can vary within a merged group.
    """

    so_sales_org: str | None       # Sales Org
    distr_chnl: str | None         # Distr Channel
    sales_office: str | None       # rendered as BRANCH
    sold_to_party: str | None      # Sold to party
    party_code: str | None         # normalized display code
    transport_mode: str | None     # from CustomerCode.transport_mode
    destination: str | None        # from CustomerCode.destination
    customer_name: str | None      # mapped customer
    stock_quantity: float          # Σ stock_quantity for the merged group


class RakeDrilldownResponse(BaseModel):
    """RAKE drill-down payload — individual jsw + jvml rows for one RAKE."""

    rake: str
    date: str
    region_id: str | None
    region_name: str
    days_filter: DaysFilter
    rows: list[RakeDrilldownRow]
    merged_rows: list[RakeDrilldownMergedRow] = []  # rows collapsed by 8-field identity
    total_quantity: float


class ReportResponse(BaseModel):
    """Full report payload — the ``data`` inside the success envelope."""

    date: str
    report_type: ReportTypeSelection  # "both" ⇒ merged jsw + jvml, grouped by SO Sales Org
    region_id: str | None
    region_name: str               # resolved name, or "All Regions"
    days_filter: DaysFilter
    ccas: list[str]               # e.g. ["VJ0H", "1000"] (jsw) / ["JV0H"] (jvml)
    has_stock: bool                # False → "No stock excel for this date selected"
    has_credit_report: bool        # False → all credit columns "NO CREDIT REPORT FOUND"
    coil_price_per_qty: float | None
    rake_columns: list[str]        # sorted unique RAKE values for this report
    rows: list[ReportPivotRow]
    grand_total: float
    grand_nco_yes_do: float
    grand_required_credit: float | None
    rake_totals: dict[str, float] = {}          # RAKE → Σ stock_quantity
    transport_mode_totals: dict[str, float] = {}  # transport_mode → Σ stock_quantity
